Posts/2021/02/XML-NewEmployees.md

```diff
@@ -0,0 +1,180 @@
+
+To create the xml, we're going to use the XmlWriter class, put this into a variable to build upon and add some formatting options. Like so:
+
+```powershell
+$XMLPath = "C:\Users\sam\AD-References.xml"
+$XMLWriter = New-Object System.XML.XMlTextWriter($XMLPath,$null)
+$xmlWriter.Formatting = 'Indented'
+$xmlWriter.Indentation = 1
+$XmlWriter.IndentChar = "`t"
+```
+If you run that, nothing crazy happens. We have a new xml source file, with nothing inside; we have set ourselves up to flesh it out.
+
+```powershell
+# write the header
+$xmlWriter.WriteStartDocument()
+$XMLWriter.WriteStartElement('Companies')
+# set XSL statements
+$xmlWriter.WriteProcessingInstruction("xml-stylesheet", "type='text/xsl' href='style.xsl'")
+```
+Okay awesome, now we're getting somewhere. We have started our "Companies" element. Let's create a company and add some office information beneath that. Then we'll finalise the XML document.
+
+```powershell
+# Start Company 1
+$XMLWriter.WriteStartElement('Company-1')
+
+# Start UK Offices
+$XMLWriter.WriteStartElement('UK-Offices')
+
+# Create Birmingham Office
+$xmlWriter.WriteStartElement('Birmingham')
+$XmlWriter.WriteElementString('Street', 'Unit 77, 132 Dummy Lane')
+$XmlWriter.WriteElementString('City', 'Birmingham')
+$XMLWriter.WriteElementString('State', 'Staffordshire')
+$XMLWriter.WriteElementString('Postcode', 'B1 1BB')
+$XMLWriter.WriteElementString('Country', 'GB')
+$XMLWriter.WriteEndElement()
+
+# End UK Offices
+$XMLWriter.WriteEndElement()
+
+#End Company 1
+$XMLWriter.WriteEndElement()
+
+# Finish up document
+$XMLWriter.WriteEndElement()
+$xmlWriter.WriteEndDocument()
+$xmlWriter.Flush()
+$xmlWriter.Close()
+```
+
+So, let's see the finished XML after we run this code.
+
+```xml
+<?xml version="1.0"?>
+<Companies>
+	<?xml-stylesheet type='text/xsl' href='style.xsl'?>
+	<Company-1>
+		<UK-Offices>
+			<Birmingham>
+				<Street>Unit 77, 132 Dummy Lane</Street>
+				<City>Birmingham</City>
+				<State>Staffordshire</State>
+				<Postcode>B1 1BB</Postcode>
+				<Country>GB</Country>
+			</Birmingham>
+		</UK-Offices>
+	</Company-1>
+</Companies>
+```
+## PowerShell - Working With XML
+
+So, now we have some useable data to work with PowerShell on. You can add more elements for Offices/Companies you are creating users for. It will become clear how you reference this next. 
+
+Tip: Save your XML-Creation script in the same location as the script referencing it; this makes for easier path variables, but also allows you to quickly create it, should the xml go missing.
+
+```powershell
+
+    # Load up the XML for reference later, if it's been deleted recreate it using the creation script.
+
+    $XMLPath = "$PSScriptRoot\AD-References.xml"
+    
+    if (Test-Path $XMLPath) {
+        
+        $XML = New-Object -TypeName XML
+        $XML.Load($XMLPath)
+    }
+    else {
+        Start-Process "$PSScriptRoot\Write-XML.ps1" -Wait
+        $XML = New-Object -TypeName XML
+        $XML.Load($XMLPath)
```

Comment:
# A simpler way?
$XML  = [XML] (Get-Content $XMLPath)

Reply:
instead of line 110? I like it! 👍